src/cmd/ksh93/sh/subshell.c

```diff
@@ -478,6 +480,7 @@ Sfio_t *sh_subshell(Shell_t *shp,Shnode_t *t, volatile int flags, int comsub)
 {
 	struct subshell sub_data;
 	register struct subshell *sp = &sub_data;
+	struct rand *rp = (struct rand*)RANDNOD->nvfun;
```

Comment:
I'm getting an intermittent crash when running the variables.sh regression tests:
```
test variables begins at 2021-05-01+19:42:39
shtests: line 375: 2594: Memory fault
test variables failed at 2021-05-01+19:42:40 with exit code 267 [ 171 tests (killed by SIGSEGV) ]
test variables(C.UTF-8) begins at 2021-05-01+19:42:40
test variables(C.UTF-8) passed at 2021-05-01+19:42:40 [ 171 tests 0 errors ]
test variables(shcomp) begins at 2021-05-01+19:42:40
shtests: line 415: 6157: Memory fault
test variables(shcomp) failed at 2021-05-01+19:42:41 with exit code 267 [ 171 tests (killed by SIGSEGV) ]
```

The `RANDNOD->nvfun` pointer can change inside of the subshell, which sometimes causes this crash (since `rp` becomes an alias to the wrong pointer).

Additionally, if the last command in a script is inside of a subshell, `$RANDOM` doesn't get reseeded (since ksh avoids creating a subshell as an optimization). This bug isn't specific to `$RANDOM`, as `${.sh.subshell}` also exhibits a similar bug:
```sh
$ arch/*/bin/ksh -c 'RANDOM=123; (echo $RANDOM)'
32365
$ arch/*/bin/ksh -c 'RANDOM=123; (echo $RANDOM)'
32365
$ arch/*/bin/ksh -c 'RANDOM=123; (echo $RANDOM)'
32365
$ arch/*/bin/ksh -c '(echo ${.sh.subshell})'
0
```

<details open>
<summary>The following patch should fix all of these issues</summary>

```diff
--- a/src/cmd/ksh93/include/variables.h
+++ b/src/cmd/ksh93/include/variables.h
@@ -34,6 +34,7 @@ struct rand
 	unsigned int	rand_seed;
 	int32_t		rand_last;
 };
+unsigned int rand_seed_seq;
 
 /* The following defines must be kept synchronous with shtab_variables[] in data/variables.c */
 
--- a/src/cmd/ksh93/sh/subshell.c
+++ b/src/cmd/ksh93/sh/subshell.c
@@ -111,7 +111,6 @@ static struct subshell
 static char subshell_noscope;	/* for temporarily disabling all virtual subshell scope creation */
 
 static unsigned int subenv;
-static unsigned int rand_seed_seq;
 
 /*
  * This routine will turn the sftmp() file into a real /tmp file or pipe
@@ -480,7 +479,7 @@ Sfio_t *sh_subshell(Shell_t *shp,Shnode_t *t, volatile int flags, int comsub)
 {
 	struct subshell sub_data;
 	register struct subshell *sp = &sub_data;
-	struct rand *rp = (struct rand*)RANDNOD->nvfun;
+	struct rand *rp;
 	int jmpval,isig,nsig=0,fatalerror=0,saveerrno=0;
 	unsigned int savecurenv = shp->curenv;
 	int savejobpgid = job.curpgid;
@@ -605,6 +604,7 @@ Sfio_t *sh_subshell(Shell_t *shp,Shnode_t *t, volatile int flags, int comsub)
 		shp->cpid = 0;
 		sh_sigreset(0);
 		/* save the current $RANDOM seed and state */
+		rp = (struct rand*)RANDNOD->nvfun;
 		sp->rand_seed = rp->rand_seed;
 		sp->rand_last = rp->rand_last;
 		/* reseed $RANDOM for the subshell */
@@ -867,6 +867,7 @@ Sfio_t *sh_subshell(Shell_t *shp,Shnode_t *t, volatile int flags, int comsub)
 		shp->cpipe[1] = sp->cpipe;
 		shp->coutpipe = sp->coutpipe;
 		/* restore $RANDOM seed and state */
+		rp = (struct rand*)RANDNOD->nvfun;
 		rp->rand_seed = sp->rand_seed;
 		rp->rand_last = sp->rand_last;
 		srand(rp->rand_seed);
--- a/src/cmd/ksh93/sh/xec.c
+++ b/src/cmd/ksh93/sh/xec.c
@@ -1937,8 +1937,10 @@ int sh_exec(register const Shnode_t *t, int flags)
 			flags &= ~OPTIMIZE_FLAG;
 			if(!shp->subshell && !shp->st.trapcom[0] && !shp->st.trap[SH_ERRTRAP] && (flags&sh_state(SH_NOFORK)))
 			{
+				/* This is the last command, so avoid creating a subshell */
 				char *savsig;
 				int nsig,jmpval;
+				struct rand *rp = (struct rand*)RANDNOD->nvfun;
 				struct checkpt *buffp = (struct checkpt*)stkalloc(shp->stk,sizeof(struct checkpt));
 				shp->st.otrapcom = 0;
 				if((nsig=shp->st.trapmax*sizeof(char*))>0 || shp->st.trapcom[0])
@@ -1948,6 +1950,12 @@ int sh_exec(register const Shnode_t *t, int flags)
 					memcpy(savsig,(char*)&shp->st.trapcom[0],nsig);
 					shp->st.otrapcom = (char**)savsig;
 				}
+				/* reseed $RANDOM */
+				rp->rand_seed = shgd->current_pid ^ ++rand_seed_seq;
+				srand(rp->rand_seed);
+				rp->rand_last = -1;
+				/* increment ${.sh.subshell} */
+				shgd->realsubshell++;
 				sh_sigreset(0);
 				sh_pushcontext(shp,buffp,SH_JMPEXIT);
 				jmpval = sigsetjmp(buffp->buff,0);
--- a/src/cmd/ksh93/tests/variables.sh
+++ b/src/cmd/ksh93/tests/variables.sh
@@ -52,6 +52,11 @@ integer rand1=$(
 integer rand2=${ print $RANDOM ;}
 (( rand1 == rand2 )) && err_exit "Test 2: \$RANDOM seed in subshell doesn't change" \
 	"(both results are $rand1)"
+# $RANDOM should be reseeded when the final command is inside of a subshell
+rand1=$($SHELL -c 'RANDOM=1; (echo $RANDOM)')
+rand2=$($SHELL -c 'RANDOM=1; (echo $RANDOM)')
+(( rand1 == rand2 )) && err_exit "Test 3: \$RANDOM seed in subshell doesn't change" \
+	"(both results are $rand1)"
 # Virtual subshells should not influence the parent shell's RANDOM sequence
 RANDOM=456
 exp="$RANDOM $RANDOM $RANDOM $RANDOM $RANDOM"
@@ -736,6 +741,12 @@ actual=$(
 expect=$'4\n3\n3\n2\n1'
 [[ $actual == "$expect" ]] || err_exit "\${.sh.subshell} failure (expected $(printf %q "$expect"), got $(printf %q "$actual"))"
 
+# ${.sh.subshell} should increment when the final command is inside of a subshell
+exp=1
+got=$($SHELL -c '(echo ${.sh.subshell})')
+[[ $exp == $got ]] || err_exit '${.sh.subshell} fails to increment when the final command is inside of a subshell' \
+	"(expected '$exp', got '$got')"
+
 unset IFS
 if	((SHOPT_BRACEPAT)) && command set -o braceexpand
 then	set -- {1..32768}
```

<details>

Reply:
I found another case where `RANDOM` needs to be reseeded:  https://github.com/ksh93/ksh/blob/1aec9b06dd0de6c6796702cd128fd796ec7b1c09/src/cmd/ksh93/sh/xec.c#L1963-L1978

Following script outputs five identical values:
```sh
( echo $RANDOM & )
( echo $RANDOM & )
( echo $RANDOM & )
( echo $RANDOM & )
( echo $RANDOM & )
:
```

I'll add a fix to your patch.

Reply:
Yikes, I just found a whopper of a bug when triggering that optimisation within a virtual subshell:
```
$ ksh -c '( ( : & ); : ); echo this gets run twice'         
this gets run twice
this gets run twice
```
Simple enough to work around with a `!shp->subshell` check.